### PR TITLE
add integrity checking into vector index commit logs

### PR DESCRIPTION
**Pull request by jeroiraz**
### What's being changed:

This PR incorporates integrity checking into vector commit logs in a backward compatible manner.

The approach taken to be backward compatible was to attach a checksum file per each commit log file. Every entry in the commit log has an associated digest into the checksum file.
At parsing time both the commit log entries and checksum entries are read in sequential fashion when recovering from disk.
The condensor process also validates integrity checking while reading commit entries into memory and when the new condensed commit log file is generated.

This approach ensures backward compatibility because if the checksum file is not present, the commit log file is read as in former versions. And older versions may ignore the presence of this checksum files (when downgrading).

Note: if an upgrade is made after a downgrade, checksum may not validate so it may be better to remove checksum first

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/vector/hnsw/startup.go [hunk 1]
```diff
@@ -88,10 +91,34 @@ func (h *hnsw) restoreFromDisk() error {
...
+			reader = commitlog.NewNoopReaderHasher(fdBuf)
+			checksumReader = commitlog.NewNoopChecksumReader()
+		}
+
 		var valid int
-		state, valid, err = NewDeserializer(h.logger).Do(fdBuf, state, false)
+		state, valid, err = NewDeserializer(h.logger).Do(reader, checksumReader, state, false)
 		if err != nil {
-			if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+			if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) || errors.Is(err, commitlog.ErrInvalidChecksum) {
```
Comment (trengrj) [inline adapters/repos/db/vector/hnsw/startup.go:121]: I think we need to recover differently on ErrInvalidChecksum. You can see below the original recovery is via `os.Truncate` to delete any extra bytes read.

Comment (jeroiraz) [reply adapters/repos/db/vector/hnsw/startup.go:121]: the intention was to still use the original truncate up to the point it was successfully read. There the truncate of the checksum file is missing...

File: adapters/repos/db/vector/hnsw/startup.go [hunk 2]
```diff
@@ -88,10 +91,34 @@ func (h *hnsw) restoreFromDisk() error {
...
-		var valid int
-		state, valid, err = NewDeserializer(h.logger).Do(fdBuf, state, false)
+		var reader rwhasher.ReaderHasher
+		var checksumReader commitlog.ChecksumReader
+
+		checksumFileName := commitLogChecksumFileName(fileName)
+		hasChecksumFile, err := fileExists(checksumFileName)
 		if err != nil {
-			if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+			return errors.Wrapf(err, "verifying checksum file existence %q", checksumFileName)
```
Comment (asdine) [inline adapters/repos/db/vector/hnsw/startup.go:100]: This logic is repeated in the memory condensor as well, should we move it inside the `NewDeserializer` function?
That would also simplify the parameters of most of `Deserializer` methods

File: adapters/repos/db/vector/hnsw/condensor_integration_test.go [hunk 3]
```diff
@@ -568,7 +569,11 @@ func TestCondensorWithoutEntrypoint(t *testing.T) {
 		require.Nil(t, err)
 
 		bufr := bufio.NewReader(fd)
-		res, _, err := NewDeserializer(logger).Do(bufr, &initialState, false)
+
+		reader := commitlog.NewNoopReaderHasher(bufr)
```
Comment (asdine) [inline adapters/repos/db/vector/hnsw/condensor_integration_test.go:573]: Not sure if I missed it but is there a test that tests the whole serialization/deserialization (not noop)?

**Review by asdine — DISMISSED** [review]
LGTM

File: adapters/repos/db/vector/hnsw/condensor_integration_test.go (hunk 3 above)
Comment (jeroiraz) [reply adapters/repos/db/vector/hnsw/condensor_integration_test.go:574]: integration tests at `condensor_integration_test.go` are now fully making use of it

File: adapters/repos/db/vector/hnsw/startup.go (hunk 2 above)
Comment (jeroiraz) [reply adapters/repos/db/vector/hnsw/startup.go:100]: managed to de-duplicate condensor code by reusing commit loggers but this initialization steps may even be more general than this steps of loading the checksum file. I got your point and will take a second look at it

**Comment by trengrj** [comment]
Hey @jeroiraz,

I ran a test import and got the following:

```
2.2M  9 Sep 13:55 .1725854136.checksum
11M  9 Sep 13:55 1725854136
```

The checksum size is larger than expected at over 10% of original data. Is this due to checksumming each operation type?

**Comment by jeroiraz** [comment]
> Hey @jeroiraz,
> 
> I ran a test import and got the following:
> 
> ```
> 2.2M  9 Sep 13:55 .1725854136.checksum
> 11M  9 Sep 13:55 1725854136
> ```
> 
> The checksum size is larger than expected at over 10% of original data. Is this due to checksumming each operation type?

Integrity checking was added with existing record types and ensuring there wouldn't be a try to allocate memory with an invalid value. Which required more than a checksum added in some record types.
Each checksum is 4bytes (crc32). 

I'd expect commit files to be eventually removed i.e. when generating snapshots

Alternatively,  checksums could be added eventually and not for each record but potentially losing more entries if there is an inconsistency...

**Review by asdine — DISMISSED** [review]

**Comment by trengrj** [comment]
@jeroiraz  I tried to simulate corruption via a couple ways:

1) corrupting the first byte of the commit log

```
load some vectors (using `go run main.go ann-benchmark -v ~/datasets/sphere-1M-meta-dpr.hdf5 -d dot` and cancelling early)
stop weaviate
dd if=/dev/urandom of=data/vector/dJcApmVO20LB/main.hnsw.commitlog.d/1728540520 bs=1 count=1 conv=notrunc
restart
```

This gave:

```
ERRO[0001] Unable to load shard dJcApmVO20LB: init shard "vector_dJcApmVO20LB": hnsw index: init index "main": restore hnsw index "main": deserialize commit log "data/vector/dJcApmVO20LB/main.hnsw.commitlog.d/1728540520": unrecognized commit type 56  build_git_commit=23e4232da build_go_version=go1.22.5 build_image_tag=localhost build_wv_version=1.27.0-alpha error="init shard \"vector_dJcApmVO20LB\": hnsw index: init index \"main\": restore hnsw index \"main\": deserialize commit log \"data/vector/dJcApmVO20LB/main.hnsw.commitlog.d/1728540520\": unrecognized commit type 56"
ERRO[0001] failed to load shard: Unable to load shard dJcApmVO20LB: init shard "vector_dJcApmVO20LB": hnsw index: init index "main": restore hnsw index "main": deserialize commit log "data/vector/dJcApmVO20LB/main.hnsw.commitlog.d/1728540520": unrecognized commit type 56  action=load_shard build_git_commit=23e4232da build_go_version=go1.22.5 build_image_tag=localhost build_wv_version=1.27.0-alpha shard_name=dJcApmVO20LB
```

2) corrupting 10MB in file

```
load vectors
stop weaviate
dd if=/dev/urandom of=data/vector/nuQ6IVXl2hNZ/main.hnsw.commitlog.d/1728540920 bs=1K count=10 seek=10M conv=notrunc
restart
```

This gave the below error, on restart once. I noticed the checksum file in this case didn't change but the original file did change. Inspecting the error I saw `EOF: reading checksum"` so looks like the checksum is working here.

```
ERRO[0001] write-ahead-log ended abruptly, some elements may not have been recovered  action=hnsw_load_commit_log_corruption build_git_commit=23e4232da build_go_version=go1.22.5 build_image_tag=localhost build_wv_version=1.27.0-alpha path=data/vector/nuQ6IVXl2hNZ/main.hnsw.commitlog.d/1728540920
```

**Pull request closed without merging**